### PR TITLE
Enable SR to rename shuttles

### PR DESCRIPTION
**Pull request by Mnemotechnician**
## About the PR
This PR adds a new field to the ID card console below the "full name" and "job title" fields, that indicates the name of the shuttle associated with the target ID. If the ID has no associated shuttle, a placeholder string is displayed. This allows the SR to rename shuttle upon demand (e.g. if someone decides to build a custom shuttle from a Construct, or simply just give their shittle a better name)

## Why / Balance
So we won't have to beg admins to rename our shuttles.

## Technical details
Added a new type of message, WriteToShuttleDeedMessage. Changed IdCardConsoleWindow (both xaml and c#) and the associated BUI, changed IdCardConsoleSystem to accept requests to modify shuttle names, added a method to ShipyardSystem that renames the provided shuttle.

## Media
# (Outdated - see the video in the conversation below!)

Screenshot_20231117_000539
Screenshot_20231117_000556
Screenshot_20231117_000608
Screenshot_20231117_000655
No owned shuttle:
Screenshot_20231117_000756

- [X] I have added screenshots/videos to this PR showcasing its changes ingame

## Breaking changes
ShuttleDeedComponent's ShuttleName field was split into ShuttleNamePrefix, ShuttleName, ShuttleNameSuffix. ShipyardSystem.Consoles is now responsible for parsing the shuttle name into these 3 parts.

**Changelog**
:cl:
- add: NanoTrasen has outfitted your local Frontier outpost with an updated ID card computer, which allows to rename shuttles! Make sure to ask your station representative.

### CODE REVIEW

**Comment by dvir001** [comment]
If you adding this you need to also make it so the name can be changed on the shipyard or its going to be hassle for SR,

Make sure you cannot change the suffix at all.

Make sure fax copy it, and warp copy it

**Comment by Mnemotechnician** [comment]
> If you adding this you need to also make it so the name can be changed on the shipyard or its going to be hassle for SR,

The point is that this should not be something that everyone can do, otherwise people will abuse this or set silly names


> Make sure you cannot change the suffix at all.

As specified above, this should be the SR's job, though I can try to make it preserve the suffix

> Make sure fax copy it, and warp copy it and; Make sure it cannot use the same names of other existing POI or ships on the map.

How do I iterate over all grids on the map? And if it's occupied, what do I show to the user? Do I leave them uninformed? Even then, you could use a name that's visually indistringuishable from the POI's name, although this will not be an issue if I make it retain suffixes (it will be "Tinna's Rest QT-378" vs "Tinna's Rest")

> Make sure the name cannot be super long.

Will do.

> Might also need to keep the original ship class so it can be recognize on the radar and just add the nickname Prefix NAME (Nickname) Suffix

This kinda nihilates the whole point, you can buy a construct and rename it to MyCorp Falcon QT-245 rather than sticking to the original name

**Comment by Cheackraze** [comment]
ambitious first PR, i like it.  There are good arguments on both sides here.  I agree that we should probably have some kind of moderating aspect to what people can name their ships. Preserving the suffix is going to be difficult without using some RegEx type coding, so I'm willing to ignore that part.  Making sure that Fax machines get renamed properly would be ideal. Warp points I think are less important, since only ghosts+admins really care about those.  And ghosts that *arent* admins have no way of seeing a changed name anyway.  Verified name sizes would be good as well as changing fax names. Other aspects I will have to look at more closely.

Otherwise, great work, I super super appreciate it, and definitely willing to help/work with you to get this in.  Dont be discouraged by our slower dev pace compared to upstream about it, and if this doesnt get the attention it deserves, do not fret.

Consider this a preliminary approval of the idea itself!

**Comment by Mnemotechnician** [comment]
Does this look better? The last field cannot be edited and is ignored on the backend. I'll fix the rest tomorrow.

image

**Comment by YumiGumi** [comment]
I love this idea, but i do have some reservations on changing the Manufacturers Prefix, changing the ship name i see as absolutely viable, but removing the manufacturers prefix kinda takes away a bit of the identity of the vessel and removes credit from who designed the ship in my eyes. We can use the SLI Gourd that i designed as an example, when someone changes the ship name, you wouldn't change the manufacturers code as the ship is still designed and sold by them, you would just change the designation the ship is built by, i would much rather see SLI <insert name> NX-### than something like a gourd with entirely unrecognizable name and title compared to at least having the manufacturer prefix. 

I will certainly say this is a heavily biased thought, and it's more I don't really wanna see someone trying to take credit for work myself and others painstakingly put in, with that being said, i heavily support seeing this kinda thing in the game, and having the ability at all to rename the ship would be a welcomed addition in my eyes.

**Comment by Mnemotechnician** [comment]
Okay, here's a new version preview (it's 20mb, github only allows uploads up to 10mb)
https://drive.google.com/file/d/1fBGyFEaJQo4u8GnMtDiYlRo57SiGrr26/view?usp=drive_link

* The name is now split into 3 parts: prefix, name, suffix (ShuttleDeedComponent was changed to reflect the change, ShipyardConsoleSystem now tries to parse shuttle name into these 3 parts upon purchase). Systems responsible for shuttle renaming accept a "newSuffix" parameter, but IdCardConsoleSystem specifically ignores it to disallow suffix modification.
* There are length constraints. IIRC, they are: up to 5 chars for prefix, up to 30 for name, and up to 7 for suffix (which is ignored, once again)
* The change is reflected everywhere (including FAXes - turns out I didn't have to change anything!)
* The change does not interfere with ship selling

**Comment by Cheackraze** [comment]
Also, we just recently dropped all the shuttle prefixes. You may need to adjust your parsing logic now. On the upside, it should be much simpler to parse now

File: Content.Server/Shipyard/Systems/ShipyardSystem.cs [hunk 1]
```diff
@@ -12,11 +12,13 @@
 using Content.Shared.CCVar;
 using Robust.Shared.Configuration;
 using System.Diagnostics.CodeAnalysis;
+using System.Linq;
 using System.Numerics;
-using Content.Shared.Coordinates;
 using Content.Shared.Shipyard.Events;
 using Content.Shared.Mobs.Components;
+using FastAccessors;
```
[the remark this answers is not in the document]
Comment (Mnemotechnician) [reply Content.Server/Shipyard/Systems/ShipyardSystem.cs:19]: Oops. This is unused, it was added by Rider when I was trying to do something there.

**Comment by Mnemotechnician** [comment]
image

**Review by Cheackraze — CHANGES_REQUESTED** [review]
since we dropped the prefixes to all the ships, it should be removed from the console and code as well

**Comment by Mnemotechnician** [comment]
image

**Comment by Cheackraze** [comment]
amazing, thank you so much for all the hard work you've done on this. Cant wait to ban the first gamer word shuttles!

**Review by Cheackraze — APPROVED** [review]

**Pull request merged**